tempesta_fw/apm.c

```diff
@@ -0,0 +1,734 @@
+		return -1;
+	return 0;
+}
+
+static int
+tfw_apm_rngmerge(TfwPcntCalc *pcalc, TfwPcntRanges **rnglst, size_t rlsz)
+{
+	int i, r, b, pcsz;
+	TfwPcntCtl pctl[TFW_STATS_RANGES];
+
+	for (i = 0, pcsz = 0; i < rlsz; ++i) {
+		spin_lock(&tfw_stats_sa_guard);
+		for (r = 0; r < TFW_STATS_RANGES; ++r)
+			pctl[r].atomic = rnglst[i]->ctl[r].atomic;
+		spin_unlock(&tfw_stats_sa_guard);
+
+		for (r = 0; r < TFW_STATS_RANGES; ++r) {
+			for (b = 0; b < TFW_STATS_BCKTS; ++b) {
+				if (!atomic_read(&rnglst[i]->cnt[r][b]))
+					continue;
+				pcalc[pcsz].intrvl = i;
+				pcalc[pcsz].range = r;
+				pcalc[pcsz].bucket = b;
+				pcalc[pcsz++].rtmval =
+					pctl[r].begin + (b << pctl[r].order);
+			}
+		}
+	}
+	sort(pcalc, pcsz, sizeof(TfwPcntCalc), tfw_apm_calc_cmp, NULL);
+
+	return pcsz;
+}
+
+static int
+tfw_apm_rnglst(TfwApmRBuf *rbuf, TfwPcntRanges **rnglst)
+{
+	TfwApmRBufEntry *rbent;
+	unsigned long jmax = 0, jtmwstart;
+	int i, imax = 0, rlsz = 0, i_rlsz;
+
+	/*
+	 * Find the entry with the maximim @jtstamp. That's the most
+	 * recent entry. The entries that precede the current entry
+	 * are used in the calculation of percentiles;
+	 */
+	for (i = 0; i < rbuf->rbentsz; ++i) {
+		if (jmax < rbuf->rbent[i].jtstamp) {
+			jmax = rbuf->rbent[i].jtstamp;
+			imax = i;
+		}
+	}
+	/*
+	 * If the entry with the maximum @jstamp is the current active
+	 * entry, then add it to the list of entries. Take care of the
+	 * circular nature of the buffer.
+	 */
+	rlsz = tfw_apm_tmwscale;
+	if (rbuf->rbent[imax].jtstamp >= jiffies - tfw_apm_jtmintrvl)
+		rlsz++;
+	imax += rbuf->rbentsz;
+	/*
+	 * Collect entries that are within the defined time window.
+	 * Most recent entries are put first in the list, older entries
+	 * are put last in the list. Take care of periods of inactivity.
+	 */
+	jtmwstart = jiffies - tfw_apm_jtmwindow;
+	for (i = 0, i_rlsz = 0; i < rlsz; ++i, --imax) {
+		rbent = &rbuf->rbent[imax % rbuf->rbentsz];
+		if (rbent->jtstamp && (rbent->jtstamp >= jtmwstart))
+			rnglst[i_rlsz++] = &rbent->pcntrng;
+	}
+
+	return i_rlsz;
+}
+
+static inline int
+__tfw_apm_calc(TfwApmData *data, Percentile *pcntl, size_t plsz, TfwPool *pool)
+{
+	int size, rlsz, pcsz;
+	TfwPcntCalc *pcalc;
+	TfwPcntRanges *rnglst[tfw_apm_tmwscale + 1];
+
+	rlsz = tfw_apm_rnglst(&data->rbuf, rnglst);
+	BUG_ON(rlsz > tfw_apm_tmwscale + 1);
+
+	size = rlsz * TFW_STATS_RANGES * TFW_STATS_BCKTS;
+	if (pool)
+		pcalc = tfw_pool_alloc(pool, size * sizeof(TfwPcntCalc));
+	else
+		pcalc = kmalloc(size * sizeof(TfwPcntCalc), GFP_KERNEL);
+	if (!pcalc)
+		return -ENOMEM;
+
+	pcsz = tfw_apm_rngmerge(pcalc, rnglst, rlsz);
+	BUG_ON(pcsz > size);
+
+	tfw_stats_calc(pcalc, pcsz, rnglst, rlsz, pcntl, plsz);
+
+	if (pool)
+		tfw_pool_free(pool, pcalc, size * sizeof(TfwPcntCalc));
+	else
+		kfree(pcalc);
+
+	return 0;
+}
+
+int
+tfw_apm_calc(void *data, Percentile *pcntl, size_t plsz, TfwPool *pool)
+{
+	BUG_ON(!data);
+	return __tfw_apm_calc((TfwApmData *)data, pcntl, plsz, pool);
```

Comment:
With [schedulers](https://github.com/tempesta-tech/tempesta/issues/565) in mind we expect high contention on statistic recalculation. So lets calculate statistic in a `HZ / 20` timer only. The timer function must save calculated statistic in one of two static buffers while `tfw_apm_calc()` will use the other buffer meantime. Since the first timer runs the timer function will find most ring buffer items zero but the current one, it still can calculate current statistics.